BackupRetention: Refactor radio retention options to be in a component state

Initsogar opened the pull request:
Related to #73035

The idea of this PR is to address multiple comments made as part of the code review in #73035. This includes a refactor in the way we are handling the `retentionOptionsCards` which will be redeclaring those on every render.

## Proposed Changes


* Move `retentionOptionsCards` to a component state as an array instead of an object and modify it dynamically when any value changes. **Addresses:** [[redeclare on every render](https://github.com/Automattic/wp-calypso/pull/73035/files#r1104532289)], [[array vs object](https://github.com/Automattic/wp-calypso/pull/73035/files#r1104531199)].
* Rename spaceNeeded prop to spaceNeededInBytes for more clarity. **Addresses:** [[odd prop name](https://github.com/Automattic/wp-calypso/pull/73035/files#r1104529533)].
* Remove unnecessary `RETENTION_OPTIONS` constant. Use RetentionPeriod type instead on `RetentionRadioOptionType` type. **Addresses:** [[simple and probably unnecessary object](https://github.com/Automattic/wp-calypso/pull/73035/files#r1104532911)], [[comment incomplete?](https://github.com/Automattic/wp-calypso/pull/73035/files#r1104519512)]
* Stop using `useMemo` on helper functions. Given that they are not really complex, and we moved the logic to a useEffect() it is preferable to not use a hook for those calculations. **Addresses:** [[are those memoizations needed?](https://github.com/Automattic/wp-calypso/pull/73035/files#r1104533766)]
* Remove unnecessary `usePrepareRetentionOptions` hook. No longer needed.
* Refactored unit tests

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### Acceptance testing
* Start a Jetpack Cloud live branch
* Navigate to Settings and wait for the new setting to load.
* The Update settings button will be disabled when the current plan or an option that requires an upgrade are selected.
* You could try to update the days of backups saved. The way to re-confirm it was properly saved is refreshing the page.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

Initsogar committed:
Move retentionOptionsCards to a component state

* Rename spaceNeeded prop to spaceNeededInBytes and make it a number type.
* Wrap `retentionOptionsCards` in a component state as an array instead of an object and modify it dynamically when any value changes.
* Remove unnecessary `RETENTION_OPTIONS` constant. Use RetentionPeriod type instead on `RetentionRadioOptionType` type.
* Remove unnecessary `usePrepareRetentionOptions` hook
* Stop using useMemo on helper functions.

Initsogar committed:
Fix unit tests. Remove unnecessary test cases.

tyxla commented on client/components/backup-retention-management/hooks.tsx:
```diff
@@ -1,24 +1,10 @@
-		[ lastBackupSize ]
-	);
+export function getEstimatedCurrentSiteSize( lastBackupSize: number ): number {
+	return lastBackupSize + lastBackupSize * STORAGE_ESTIMATION_ADDITIONAL_BUFFER;
```
Might as well be:

```suggestion
	return lastBackupSize * ( STORAGE_ESTIMATION_ADDITIONAL_BUFFER + 1 );
```

Initsogar replied:
Yeah, it looks cleaner. Just changed it :)

tyxla commented on client/components/backup-retention-management/hooks.tsx:
```diff
@@ -28,50 +14,9 @@ export function useEstimatedCurrentSiteSize(): number {
  * @returns Space needed in bytes
  */
-export function useEstimateSpaceNeeded(
+export function getSpaceNeededInBytes(
```
Do we really need a function that multiplies 2 numbers? I think we can just replace it with a simple multiplication where necessary, can't we?

Initsogar replied:
We were anticipating to future changes, but as you said previously, it is better to follow YAGNI principle. If we needed in the future, then we will create those functions again.

Initsogar replied:
We just removed both functions and its tests :)

tyxla commented on client/components/backup-retention-management/test/hooks.tsx:
```diff
@@ -1,80 +1,25 @@
 	} );
 
-	describe( 'useEstimateSpaceNeeded', () => {
+	describe( 'getSpaceNeededInBytes', () => {
```
These tests are an indication that we don't really need this function and we can just inline the multiplication.

tyxla commented on client/components/backup-retention-management/hooks.tsx:
```diff
@@ -1,24 +1,10 @@
-		() => lastBackupSize + lastBackupSize * STORAGE_ESTIMATION_ADDITIONAL_BUFFER,
-		[ lastBackupSize ]
-	);
+export function getEstimatedCurrentSiteSize( lastBackupSize: number ): number {
```
I think we can remove this function, too. Seems like a straightforward calculation we can do inline.

Initsogar replied:
Done!

tyxla commented on client/components/backup-retention-management/index.tsx:
```diff
@@ -70,53 +73,104 @@ const BackupRetentionManagement: FunctionComponent = () => {
-	};
+	const [ retentionOptionsCards, setRetentionOptionsCards ] = useState<
+		RetentionRadioOptionType[]
+	>( [
+		{
+			label: translate( '7 days' ),
+			spaceNeededInBytes: 0,
+			upgradeRequired: false,
+			isCurrentPlan: false,
+			value: 7,
+			checked: false,
+		},
+		{
+			label: translate( '30 days' ),
+			spaceNeededInBytes: 0,
+			upgradeRequired: false,
+			isCurrentPlan: false,
+			value: 30,
+			checked: false,
+		},
+		{
+			label: translate( '120 days' ),
+			spaceNeededInBytes: 0,
+			upgradeRequired: false,
+			isCurrentPlan: false,
+			value: 120,
+			checked: false,
+		},
+		{
+			label: translate( '1 year' ),
+			spaceNeededInBytes: 0,
+			upgradeRequired: false,
+			isCurrentPlan: false,
+			value: 365,
+			checked: false,
+		},
+	] );
```
This object will be reinitialized on every re-render. We could benefit from use lazy initialization. See [this](https://kentcdodds.com/blog/use-state-lazy-initialization-and-function-updates) for more info.

It also feels like a weird data type to live in state. Let's try to refactor and use as little state as possible. The rest that is always the same can live outside of the state.

Initsogar replied:
> This object will be reinitialized on every re-render. We could benefit from use lazy initialization. See [this](https://kentcdodds.com/blog/use-state-lazy-initialization-and-function-updates) for more info.

Great resource. Thanks for the tip. I just updated this to add lazy initialization.

> It also feels like a weird data type to live in state. Let's try to refactor and use as little state as possible. The rest that is always the same can live outside of the state.

We double-checked the code and we were able to decouple some props that are not necessarily needed on the state.
* Moved `label` to a constant and use it directly when rendering the radio button.
* Handled `checked` and `isCurrentPlan` on the radio control.

tyxla reviewed (commented):
I love seeing all the red lines 🔴 ! Thanks!

I've left a few more suggestions for simplification. Let me know what you think!

Initsogar committed:
Remove unnecessary function and tests

This moves the `getEstimatedCurrentSiteSize` and `getSpaceNeededInBytes` functions to be executed in-line given that they are basic math operations and just used once. Also, we are getting rid of the tests related to those.

Initsogar committed:
Refactor `retentionOptionsCards`

* Move `label` to a constant and use it directly when rendering the radio button.
* Handle `checked` and `isCurrentPlan` on the radio control.

Initsogar committed:
Add lazy initialization to `retentionOptionsCards`

Initsogar commented:
@tyxla  I think this is ready for a re-review. It would be great if you could take another look :)

tyxla commented on client/components/backup-retention-management/constants.ts:
```diff
@@ -1,11 +1,15 @@
 export const STORAGE_RETENTION_LEARN_MORE_LINK =
 	'https://jetpack.com/support/backup/#how-is-storage-usage-calculated';
+
+export const RETENTION_OPTIONS_LABELS: Record< RetentionPeriod, TranslateResult > = {
+	[ 7 ]: translate( '7 days' ),
+	[ 30 ]: translate( '30 days' ),
+	[ 120 ]: translate( '120 days' ),
+	[ 365 ]: translate( '1 year' ),
+};
```
This way of declaring the labels once at runtime during the first import may cause some i18n issues. If the locale gets changed, or new translations get loaded in the meantime AFTER this constant has been declared, the strings will remain unchanged, and `translate()` will not be called again. 

One way forward would be to just avoid declaring these strings separately in an object, and use them inline as needed. That would be my preferred way - I don't see a good reason to reuse those strings. What do we earn by reusing them?

If we really insist these strings to be reused, we can have them in a hook like `useStorageStatusText()` [does](https://github.com/Automattic/wp-calypso/blob/9f9d18171543a4bc366b0edd2ade4631bedf57d9/client/components/backup-storage-space/usage-warning/use-storage-status-text.ts#L10), or (not recommended) use [getters](https://github.com/Automattic/wp-calypso/blob/try/upgrade-wordpress-packages-pre-react18/client/components/domains/use-my-domain/utilities/option-info.js#L39) or [callbacks](https://github.com/Automattic/wp-calypso/blob/1740a4c11f8e8fcc313e04d84e636ceacc8c4aae/client/lib/plans/features-list.tsx#L851).

Initsogar replied:
Makes totally sense! We are using this constant only in `backup-retention-management/retention-options/retention-option-card.tsx` so I think we can just put it there where we need to render them.

Initsogar replied:
Done!

tyxla commented on client/components/backup-retention-management/index.tsx:
```diff
@@ -70,53 +73,82 @@ const BackupRetentionManagement: FunctionComponent = () => {
+
+		setRetentionOptionsCards( ( previousOptions ) => {
+			const newOptions = previousOptions.map( ( option ) => {
+				const spaceNeededInBytes = estimatedCurrentSiteSize * option.id;
```
Looks like the formula for calculating `spaceNeededInBytes` is predictable: it's always equal to the ID, times `estimatedCurrentSiteSize`. That means that `spaceNeededInBytes` is likely not necessary to be in the state and can be derived from the `id`.

tyxla commented on client/components/backup-retention-management/index.tsx:
```diff
@@ -70,53 +73,82 @@ const BackupRetentionManagement: FunctionComponent = () => {
+		setRetentionOptionsCards( ( previousOptions ) => {
+			const newOptions = previousOptions.map( ( option ) => {
+				const spaceNeededInBytes = estimatedCurrentSiteSize * option.id;
+				const upgradeRequired = spaceNeededInBytes > storageLimitBytes;
```
Same as above for `upgradeRequired` - it can be derived from `spaceNeededInBytes` and `storageLimitBytes` therefore it does not need to be stored in `state`. 

There's an interesting parallel we can draw between that kind of functional component derived state and the original React derived state that is supported for the class component syntax. You can read more about `getDerivedStateFromProps()` [here](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops) and why it's usually not necessary [in this article](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html).

Initsogar replied:
Hmm, based on this, then we actually don't need a state at all.

We can replace the retentionOptionsCards state with:

On `constants.tsx`:
```javascript
export const RETENTION_OPTIONS = [ 7, 30, 120, 365 ] as RetentionPeriod[];
```

On `BackupRetentionManagement`:
```javascript
const retentionOptionsCards = RETENTION_OPTIONS.map( ( retentionDays ) => {
	return {
		id: retentionDays,
		spaceNeededInBytes: estimatedCurrentSiteSize * retentionDays,
		upgradeRequired: estimatedCurrentSiteSize * retentionDays > storageLimitBytes,
	};
} );
```

Way to simpler and we can get rid of unnecessary useEffect(s).

What do you think?

tyxla replied:
I'm really glad you got to this conclusion! Less state (and effects that maintain it) is always a good idea 👍

tyxla commented on client/components/backup-retention-management/index.tsx:
```diff
@@ -70,53 +73,82 @@ const BackupRetentionManagement: FunctionComponent = () => {
+			if ( value ) {
+				setRetentionSelected( value );
+				const selectedOption = retentionOptionsCards.find( ( option ) => option.id === value );
+				setStorageUpgradeRequired( selectedOption?.upgradeRequired ?? false );
```
Should we call `setStorageUpgradeRequired` only if the new `storageUpgradeRequired` value is different from the current one?

Initsogar replied:
You are right! Done!

tyxla commented on client/components/backup-retention-management/index.tsx:
```diff
@@ -70,53 +73,82 @@ const BackupRetentionManagement: FunctionComponent = () => {
+	const [ retentionOptionsCards, setRetentionOptionsCards ] = useState( getOptionsInitialState );
+
+	// Update the retention options cards with the correct values dynamically
+	useEffect( () => {
```
This effect seems to be unnecessary. In the subsequent comments, I'm suggesting the removal of some state properties, and after removing them, the options become independent and unchanged by this effect. 

If we really need to render something different while loading, we can instead render a placeholder while `isFetching()` is `true` and then render the options as they are. See comments below.

Initsogar replied:
We just get rid of that useEffect with the mentioned approach [here](https://github.com/Automattic/wp-calypso/pull/73303#discussion_r1107153191), so marking this as solved.

tyxla commented on client/components/backup-retention-management/index.tsx:
```diff
@@ -70,53 +73,82 @@ const BackupRetentionManagement: FunctionComponent = () => {
+			if ( value ) {
+				setRetentionSelected( value );
+				const selectedOption = retentionOptionsCards.find( ( option ) => option.id === value );
+				setStorageUpgradeRequired( selectedOption?.upgradeRequired ?? false );
```
Is there really a possibility that `selectedOption` will be nullish? Is the optional chaining necessary to satisfy TS? If yes, and `selectedOption` will always be there, perhaps we need to tell the TS compiler about that fact?

Initsogar replied:
Yeah, I got carried away by TS in this one. Just fixed it :)

tyxla commented on client/components/backup-retention-management/index.tsx:
```diff
@@ -70,53 +73,82 @@ const BackupRetentionManagement: FunctionComponent = () => {
-	}, [] );
+	const onRetentionSelectionChange = useCallback(
+		( value: number ) => {
+			if ( value ) {
```
In which case is `value` fasly? Are we intentionally not doing anything in that case?

Initsogar replied:
Double-checking noticed that we added this check because we were setting the default selection to 0. Fixing it to set a valid option as default.

Initsogar replied:
It appears that just adding a valid option as default is not too straightforward and, for timely reasons, I think it would be better to address this improvement in another PR. I commit to it. Would it be okay?

tyxla replied:
Sounds good, thanks 👍

tyxla commented on client/components/backup-retention-management/retention-options/retention-option-card.tsx:
```diff
@@ -52,7 +52,7 @@ const RetentionOptionCard: FunctionComponent< RetentionOptionCardProps > = ( {
 			<div className="retention-option__space-needed">
 				<div className="space-needed__label">{ translate( 'Space needed:' ) }</div>
-				<div className="space-needed__value">{ spaceNeeded }</div>
+				<div className="space-needed__value">{ useStorageText( spaceNeededInBytes ) }</div>
```
You might want to move the hook call to the top of the function. Not a requirement, but it gives better visibility on the fact that we're calling a hook and reduces the chance of someone breaking the [rules of hooks](https://reactjs.org/docs/hooks-rules.html) in the future.

Initsogar replied:
Done

tyxla commented on client/components/backup-retention-management/retention-options/retention-options-control.tsx:
```diff
@@ -28,10 +30,14 @@ const RetentionOptionsControl: React.FC< RetentionOptionsControlProps > = ( {
+					upgradeRequired={ option.upgradeRequired }
+					checked={ retentionSelected === option.id }
+					isCurrentPlan={ currentRetentionPlan === option.id }
+					key={ option.id }
```
I'd like to point out that using the value as a key might have unexpected consequences and cause unexpected behavior. 

That can mean that if the value changes, the key will also change, and React will be unable to reuse the component, creating a new one. It can also mean that React might unintentionally reuse an old component that had the same value (respectively, key) as the one we're now about to render.

Initsogar replied:
Given that for this specific scenario, the `id` is the unique identifier for that option (which only can be 7, 30, 120 or 365 for now), I don't see how this could be an issue. What do you think?

tyxla replied:
I don't see it problematic in this particular scenario, but it is a red flag because someone might borrow this code later and use it with a larger amount of (or infinite) options. It partially defies the idea of how React keys are supposed to serve.

That being said, feel free to add a prefix to the `key` so it's at least different from the value. Adding a comment that this is fine because we're working with a small set of option IDs could also help our future selves.

Initsogar replied:
Done!

Initsogar committed:
Move useStorageText call to top level

Initsogar committed:
Move options labels to retention-option-card

Initsogar committed:
Add a prefix to RetentionOptionCard key

Given that we are working with a small set of options, we could use the option id as a key.

Initsogar committed:
Refactor retentionOptions card to not use state

Given that most of properties could be derivaded state, we actually don’t need at state at all. This commit is to move the retentionOptions to a constant instead of state.

Initsogar committed:
Call setStorageUpgradeRequired only when needed

Initsogar committed:
Merge branch 'trunk' of https://github.com/Automattic/wp-calypso into update/backup-retention-options-to-state

Initsogar commented:
@tyxla Thanks for the detailed review. Great learning opportunity and ensuring code quality.
I think this is ready to re-review. Could you take another look?

tyxla commented on client/components/backup-retention-management/index.tsx:
```diff
@@ -65,53 +68,48 @@ const BackupRetentionManagement: FunctionComponent = () => {
 	const currentSiteSizeText = useStorageText( estimatedCurrentSiteSize );
 	const storageLimitText = useStorageText( storageLimitBytes );
 
-	const retentionOptionsCards: Record< number, RetentionRadioOptionType > = {
-		[ RETENTION_OPTIONS.RETENTION_DAYS_7 ]: usePrepareRetentionOptions(
-			translate( '7 days' ),
-			RETENTION_OPTIONS.RETENTION_DAYS_7,
-			currentRetentionPlan,
-			retentionSelected === RETENTION_OPTIONS.RETENTION_DAYS_7
-		),
-		[ RETENTION_OPTIONS.RETENTION_DAYS_30 ]: usePrepareRetentionOptions(
-			translate( '30 days' ),
-			RETENTION_OPTIONS.RETENTION_DAYS_30,
-			currentRetentionPlan,
-			retentionSelected === RETENTION_OPTIONS.RETENTION_DAYS_30
-		),
-		[ RETENTION_OPTIONS.RETENTION_DAYS_120 ]: usePrepareRetentionOptions(
-			translate( '120 days' ),
-			RETENTION_OPTIONS.RETENTION_DAYS_120,
-			currentRetentionPlan,
-			retentionSelected === RETENTION_OPTIONS.RETENTION_DAYS_120
-		),
-		[ RETENTION_OPTIONS.RETENTION_DAYS_365 ]: usePrepareRetentionOptions(
-			translate( '1 year' ),
-			RETENTION_OPTIONS.RETENTION_DAYS_365,
-			currentRetentionPlan,
-			retentionSelected === RETENTION_OPTIONS.RETENTION_DAYS_365
-		),
-	};
```
😍

tyxla reviewed (approved):
Nice work on the code, looks great 👍 🚀 

I haven't tested, so I'll leave the final approval to your team.
